server-spi-private/src/main/java/org/keycloak/models/utils/KeycloakModelUtils.java

```diff
@@ -266,6 +269,80 @@ public static void runJobInTransaction(KeycloakSessionFactory factory, KeycloakS
+     * @param attemptsCount the maximum number of attempts to execute the callable.
+     * @param retryIntervalMillis the base interval value in millis used to compute the delay.
+     * @param <V> the type returned by the callable.
+     * @return the value computed by the callable.
+     */
+    public static <V> V runJobInRetriableTransaction(final KeycloakSessionFactory factory, final KeycloakSessionTaskWithResult<V> callable,
+                                                     final int attemptsCount, final int retryIntervalMillis) {
+        int retryCount = 0;
+        Random rand = new Random();
+        V result;
+        while (true) {
+            KeycloakSession session = factory.create();
+            KeycloakTransaction tx = session.getTransactionManager();
+            try {
+                tx.begin();
+                result = callable.run(session);
+                if (tx.isActive()) {
+                    if (tx.getRollbackOnly()) {
+                        tx.rollback();
+                    } else {
+                        tx.commit();
+                    }
+                }
+                break;
+            } catch (RuntimeException re) {
+                if (tx.isActive()) {
+                    tx.rollback();
+                }
+                if (isExceptionRetriable(re) && ++retryCount < attemptsCount) {
+                    int delay = (int)(Math.pow(2, retryCount) * retryIntervalMillis) + rand.nextInt(retryIntervalMillis);
+                    try {
+                        Thread.sleep(delay);
+                    } catch (InterruptedException ie) {
+                        ie.addSuppressed(re);
+                        throw new RuntimeException(ie);
+                    }
+                } else {
+                    throw re;
+                }
+            } finally {
+                session.close();
+            }
+        }
+        return result;
+    }
+
+    /**
+     * Checks if the specified exception is retriable or not. A retriable exception must be an instance of {@code SQLException}
+     * and must have a retriable state (for example, "40001" on CockroachDB).
+     *
+     * @param exception the exception to be checked.
+     * @return {@code true} if the exception is retriable; {@code false} otherwise.
+     */
+    public static boolean isExceptionRetriable(final Exception exception) {
+        Objects.requireNonNull(exception);
+        // first find the root cause and check if it is a SQLException
+        Throwable rootCause = exception;
+        while (rootCause.getCause() != null && rootCause.getCause() != rootCause) {
+            rootCause = rootCause.getCause();
+        }
+        if (rootCause instanceof SQLException) {
+            // check if the exception state is a recoverable one (40001)
+            return "40001".equals(((SQLException) rootCause).getSQLState());
+        }
+        return false;
+    }
```

Comment:
Adding a note that I agree that we might revisit this code in the future but is ok as is now.
Reason is that 40001 is a standard SQL state as defined in SQL standard, and across the implementations its meaning boils down to "deadlock" (applies to Postgres, MSSQL, Oracle, MySQL, and others)

Reply:
Good to know that 40001 is a standard, didn't know about that. With this information, a follow up issue is IMHO not needed now, but only only when we need that extra flexibility.

It would be great to have the information from Hynek how 40001 works reflected in the JavaDoc / comments for this method.

Reply:
I've added what Hynek wrote in the Javadoc for this method, for future reference.

Reply:
ok, resolved for me.